File: python/data_structures/linked_list.py
"""
Singly Linked List
Time Complexity:
- Access/Search: O(n)
- Insertion/Deletion (Head): O(1)
Space Complexity: O(n)

Description:
A linear data structure where elements are not stored in contiguous memory.
Each node contains a value and a pointer to the next node.
"""

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

    def append(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
        self.size += 1

    def to_list(self):
        result = []
        current = self.head
        while current:
            result.append(current.value)
            current = current.next
        return result

if __name__ == "__main__":
    ll = LinkedList()
    ll.append(1)
    ll.append(2)
    ll.append(3)
    assert ll.to_list() == [1, 2, 3]
    print("Python Linked List: All tests passed!")

File: java/data_structures/LinkedList.java
package data_structures;

import java.util.ArrayList;
import java.util.List;

/**
 * Singly Linked List
 * Time Complexity:
 * - Access/Search: O(n)
 * - Insertion/Deletion (Head): O(1)
 * Space Complexity: O(n)
 */
public class LinkedList<T> {

    private static class Node<T> {
        T value;
        Node<T> next;

        Node(T value) {
            this.value = value;
        }
    }

    private Node<T> head;
    private int size;

    public void append(T value) {
        Node<T> newNode = new Node<>(value);
        if (head == null) {
            head = newNode;
        } else {
            Node<T> current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
        size++;
    }

    public List<T> toList() {
        List<T> list = new ArrayList<>();
        Node<T> current = head;
        while (current != null) {
            list.add(current.value);
            current = current.next;
        }
        return list;
    }

    public static void main(String[] args) {
        LinkedList<Integer> ll = new LinkedList<>();
        ll.append(1);
        ll.append(2);
        ll.append(3);
        assert ll.toList().equals(List.of(1, 2, 3));
        System.out.println("Java Linked List: All tests passed!");
    }
}
